teps/0021-results-api.md

```diff
@@ -0,0 +1,772 @@
+```
+
+We will not enforce a particular filter spec for the API. It will be up to API
+implementers to define and handle filters.
+
+#### Deletion of Completed Resources
+
+Since the Pipeline API is currently how most users interact with execution
+results, this means that Task/PipelineRuns need to remain persisted in etcd.
+This comes at a tradeoff, since the more objects there are in etcd results in
+the Pipeline controller needing to handle more data at execution time, which can
+have a noticable impact on Pipeline API and scheduling latency
+(https://github.com/tektoncd/pipeline/issues/1302). Users address this by
+periodicly deleting completed resources from their cluster
+(https://github.com/tektoncd/plumbing/issues/439), but this at the cost of
+losing the execution history. Offloading long-term result storage to a separate
+system will enable us to add support to clean up these completed resources to
+help reduce the load on the Pipeline controller, without sacrificing historical
+results.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Pipelines Source of Truth
+
+The Results API is intended for long term result storage for users - it is not
+meant to be a data plane replacement for the Pipeline controller. There should
+be an expectation of eventual consistency - clients that are sensitive to this
+should continue to rely on the Pipeline API directly.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+
+```
+<<[UNRESOLVED wlynch ]>>
+This section will be filled in based on feedback
+<<[/UNRESOLVED]>>
+```
+
+### User Experience (optional)
+
+<!--
+Consideration about the user experience. Depending on the area of change,
+users may be task and pipeline editors, they may trigger task and pipeline
+runs or they may be responsible for monitoring the execution of runs,
+via CLI, dashboard or a monitoring system.
+
+Consider including folks that also work on CLI and dashboard.
+-->
+
+#### Result Resources
+
+Users typically don't make large distinction between PipelineRuns or multiple
+TaskRuns that ran as part of a PipelineRun. Generally when looking at a
+dashboard, the most relevant information is what execution events happened.
+Results are intended to be this abstraction over execution types (i.e.
+PipelineRun, TaskRun) so that we can group this data for users and provide a
+mechanism to include additional metadata that does not fit neatly into TaskRuns
+or PipelineRuns.
```

Comment:
i would think we'd still store the individual TaskRuns along with the PipelineRun tho right?